text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle, always invalid.
+    fn dangling() -> Self::Handle;
+
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
+
+    //  The following methods are similar to `Allocator`, reformulated in terms of `Self::Handle`.
+
+    /// Allocates a new handle to a block of memory. On success, invalidates any existing handle.
+    fn allocate(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Deallocates a handle.
+    unsafe fn deallocate(&self, handle: Self::Handle, layout: Layout);
+
+    /// Grows the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn grow(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Shrinks the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn shrink(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Allocates a new handle to a block of zeroed memory. On success, invalidates any existing handle.
+    fn allocate_zeroed(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+
+    /// Grows the block of memory associated to a handle with zeroed memory. On success, the handle is invalidated.
+    unsafe fn grow_zeroed(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+}
+```
+
+_Note:  full-featured documentation of the trait and methods can be found in the companion repository at_
+        https://github.com/matthieu-m/store/blob/main/src/interface.rs.
+
+The `Store` trait is supplemented by 3 additional marker traits, providing extra guarantees:
+
+```rust
+/// A refinement of `Store` which does not invalidate existing handles on allocation, and does not invalidate
+/// unrelated existing handles on deallocation.
+pub unsafe trait MultipleStore: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers on allocation, resolution, or deallocation, but
+/// may invalidate them on moves.
+pun unsafe trait StableStore: Store {}
```

Comment:
```suggestion
pub unsafe trait StableStore: Store {}
```

Reply:
I like `pun`, maybe we should keep it :) ?